### PR TITLE
First approach to operator system tests

### LINKED ISSUE #1617: [System Tests] Kroxylicious operator installation
**Is your feature request related to a problem? Please describe.**
Kroxylicious operator is under development and we would need to start creating simple system tests to check the installation is going fine.

**Describe the solution you'd like**
We would follow the `SetUpClusterOperator` approach done in [Strimzi](https://github.com/strimzi/strimzi-kafka-operator/blob/main/systemtest/src/main/java/io/strimzi/systemtest/resources/operator/SetupClusterOperator.java) project.

Currently we will focus our tests only on the `bundle` installation (yaml files), but they can be done via `olm` or `helm` if required in the future.

Our class could be called `SetUpKroxyliciousOperator` and may be simplified to our current functionality (without `olm` and `helm`) and can be upgraded in the future if needed. We should start always simple.

### PR DESCRIPTION
**Pull request by franvila**
### Type of change

- Enhancement / new feature

### Description

As Kroxylicious operator is landed, we need to adapt the kroxylicious-system tests framework to be able to test the operator.

Closes #1617 

### Additional Context

* In order to make use of the same base test framework in all projects, we have started to use [TestFrame](https://github.com/skodjob/test-frame) framework that collects some common utils that makes easy the implementation of our framework.
* We are covering only **Bundle Yaml** installation. Many resources has been added to manage the installation files.
* Currently there are no oficial system tests. I have only implemented the framework and a simple test to install the operator. 
* The `OperatorST` is disabled until we have a developer image on quay.io.

### Checklist

- [X] Write tests
- [x] Update documentation
- [x] Make sure all unit/integration tests pass
- [x] Make sure all Sonarcloud warnings are addressed or are justifiably ignored.
- [x] If applicable to the change, [trigger the system test suite](../blob/main/DEV_GUIDE.md#jenkins-pipeline-for-system-tests).  Make sure tests pass.
- [x] If applicable to the change, [trigger the performance test suite](../blob/main/PERFORMANCE.md#jenkins-pipeline-for-performance). Ensure that any degradations to performance numbers are understood and justified.
- [x] Ensure the PR references relevant issue(s) so they are closed on merging.
- [x] For user facing changes, update CHANGELOG.md (remember to include changes affecting the API of the test artefacts too).

> **_NOTE:_**  You must be a member of `@kroxylicious/developers` to trigger the system test and performance test suites.  If you are not part of this group, comment on the PR requesting a trigger, tagging `@kroxylicious/developers`.

### CODE REVIEW

**Comment by franvila** [comment]
@strimzi-ci run system tests

**Comment by franvila** [comment]
@strimzi-ci run system tests

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Environment.java [hunk 1]
```diff
@@ -95,57 +111,55 @@ private Environment() {
...
     /**
      * KROXY_IMAGE_REPO env variable assignment
      */
-    public static final String KROXY_IMAGE_REPO = getOrDefault(KROXY_IMAGE_REPO_ENV, KROXY_IMAGE_REPO_DEFAULT);
+    public static final String KROXY_IMAGE_REPO = ENVIRONMENT_VARIABLES.getOrDefault(KROXY_IMAGE_REPO_ENV, KROXY_IMAGE_REPO_DEFAULT);
     /**
      * SKIP_TEARDOWN env variable assignment.
      */
-    public static final boolean SKIP_TEARDOWN = Boolean.parseBoolean(getOrDefault(SKIP_TEARDOWN_ENV, SKIP_TEARDOWN_DEFAULT));
+    public static final boolean SKIP_TEARDOWN = Boolean.parseBoolean(ENVIRONMENT_VARIABLES.getOrDefault(SKIP_TEARDOWN_ENV, SKIP_TEARDOWN_DEFAULT));
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/Environment.java:128]: Presumably,  you could express this more succinctly as:

```
    private static final boolean SKIP_TEARDOWN_DEFAULT = false;
    public static final boolean SKIP_TEARDOWN = ENVIRONMENT_VARIABLES.getOrDefault(SKIP_TEARDOWN_ENV, Boolean::parseBoolean, SKIP_TEARDOWN_DEFAULT);
```

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/enums/DeploymentType.java [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.systemtests.enums;
+
+public enum DeploymentType {
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/enums/DeploymentType.java:9]: I notice that io.skodjob.testframe.enums has InstallType.  What's the difference between this enum and that one?

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/executor/Exec.java [hunk 3]
```diff
@@ -401,8 +404,8 @@ private void storeOutputsToFile() {
         if (logPath != null) {
             try {
                 Files.createDirectories(logPath);
-                Files.writeString(Paths.get(logPath.toString(), "stdOutput.log"), stdOut, Charset.defaultCharset());
-                Files.writeString(Paths.get(logPath.toString(), "stdError.log"), stdErr, Charset.defaultCharset());
+                ReadWriteUtils.writeFile(Paths.get(logPath.toString(), "stdOutput.log"), stdOut);
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/executor/Exec.java:407]: I'm not really sure I see the value of this change.

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/installation/kroxylicious/KroxyliciousOperator.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+package io.kroxylicious.systemtests.installation.kroxylicious;
+
+import io.kroxylicious.systemtests.resources.operator.SetUpKroxyliciousOperator;
+
+/**
+ * The type Kroxylicious operator.
+ */
+public class KroxyliciousOperator {
+    private final String deploymentNamespace;
+    private SetUpKroxyliciousOperator setUpKroxyliciousOperator;
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/installation/kroxylicious/KroxyliciousOperator.java:16]: I'm not understanding why SetUpKroxyliciousOperator is a separate class.   KroxyliciousOperator with methods like deploy and undeploy makes some sense, but why delegate to another class?

**Comment by k-wall** [comment]
I'm afraid you've made the life of the reviewer really hard.  This PR is really big! It feels like every file is touched.  I'm one hour in to the PR and my head is spinning. 

PRs should be as tightly narrowly focused as possible.   If you feel the motivation to refactor things, that's great (indeed encouraged!), but it is much better to refactor as separate PR(s).   As a reviewer you need to build a mental model of what's changing - the bigger the change, the more mental energy it takes (and it is not a linear relationship).

File: kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/resources/operator/KroxyliciousOperatorBundleInstaller.java [hunk 5]
```diff
@@ -0,0 +1,483 @@
...
+            }
+        }
+    }
+
+    /**
+     * Perform application of ServiceAccount, Roles and CRDs needed for proper cluster operator deployment.
+     * Configuration files are loaded from kroxylicious-operator directory.
+     */
+    public void applyClusterOperatorInstallFiles(String namespaceName) {
+        List<File> operatorFiles = Arrays.stream(Objects.requireNonNull(new File(Constants.PATH_TO_OPERATOR_INSTALL_FILES).listFiles())).sorted()
```
Comment (k-wall) [inline kroxylicious-systemtests/src/main/java/io/kroxylicious/systemtests/resources/operator/KroxyliciousOperatorBundleInstaller.java:157]: Fabric8 lets you do the equivalent of an kubectl apply.   You should apply the CRDs first, then the rest of the resources.   You don't need all this case by case handling.

https://github.com/fabric8io/kubernetes-client/blob/main/kubernetes-examples/src/main/java/io/fabric8/kubernetes/examples/kubectl/equivalents/CreateOrReplaceResourceList.java

**Comment by franvila** [comment]
> I'm afraid you've made the life of the reviewer really hard. This PR is really big! It feels like every file is touched. I'm one hour in to the PR and my head is spinning.
> 
> PRs should be as tightly narrowly focused as possible. If you feel the motivation to refactor things, that's great (indeed encouraged!), but it is much better to refactor as separate PR(s). As a reviewer you need to build a mental model of what's changing - the bigger the change, the more mental energy it takes (and it is not a linear relationship).

Yes, you are totally right. I'm really sorry for that. I needed to touch many files and when you enter in a file you haven't touch since ages you realize that you can improve it

**Comment by k-wall** [comment]
I wanted to use the work that @robobario did on PR #1635 as an example to follow.   There what started off a single bug fix PR, grew legs and turned into a monster.   Rob made the effort to split the work into three PRs which were reviewed and merged in sequence.   I sure this was more work for Rob, but less collective effort for the team as a whole.

So with this PR, my suggestion would be to try to break it up into cohesive chunks.   Separately, I think it would be also good to talk about the motivations for the introduction of the Test Frame library and the value you see it bringing (maybe a slack thread or as a github discussion).     Code re-use is usually good thing, but there is a tension, more dependencies = more complexity.

**Comment by franvila** [comment]
Moving to #1690

**Pull request closed without merging**